src/Bicep.Core/Semantics/FunctionOverload.cs

```diff
@@ -4,29 +4,51 @@
 using System.Collections.Generic;
 using System.Collections.Immutable;
 using System.Linq;
+using Azure.Deployments.Expression.Expressions;
+using Bicep.Core.Diagnostics;
+using Bicep.Core.Emit;
 using Bicep.Core.Extensions;
 using Bicep.Core.Syntax;
 using Bicep.Core.TypeSystem;
+using Newtonsoft.Json;
 
 namespace Bicep.Core.Semantics
 {
     public class FunctionOverload
     {
         public delegate TypeSymbol ReturnTypeBuilderDelegate(IEnumerable<FunctionArgumentSyntax> arguments);
+        public delegate TypeSymbol AdvancedReturnTypeBuilderDelegate(AdvancedReturnTypeBuilderContext context);
+        public delegate bool ExpressionEmitterDelegate(JsonTextWriter writer, EmitterContext emitterContext, FunctionCallSyntax functionCallSyntax, FunctionSymbol functionSymbol, out SyntaxBase? expression, out LanguageExpression? languageExpression);
 
-        public FunctionOverload(string name, string description, ReturnTypeBuilderDelegate returnTypeBuilder, TypeSymbol returnType, IEnumerable<FixedFunctionParameter> fixedParameters, VariableFunctionParameter? variableParameter, FunctionFlags flags = FunctionFlags.Default)
+        public class AdvancedReturnTypeBuilderContext
         {
-            this.Name = name;
-            this.Description = description;
-            this.ReturnTypeBuilder = returnTypeBuilder;
-            this.FixedParameters = fixedParameters.ToImmutableArray();
-            this.VariableParameter = variableParameter;
-            this.Flags = flags;
-
-            this.MinimumArgumentCount = this.FixedParameters.Count(fp => fp.Required) + (this.VariableParameter?.MinimumCount ?? 0);
-            this.MaximumArgumentCount = this.VariableParameter == null ? this.FixedParameters.Length : (int?)null;
+            public IBinder Binder { get; }
+            public IDiagnosticWriter Diagnostics { get; }
+            public (FunctionArgumentSyntax syntax, TypeSymbol)[] Arguments { get; }
+
+            public AdvancedReturnTypeBuilderContext(IBinder binder, IDiagnosticWriter diagnostic, (FunctionArgumentSyntax syntax, TypeSymbol)[] arguments)
```

Comment:
>AdvancedReturnTypeBuilderContext [](http://example.com/codeflow?start=19&length=32)

Why not make this the parameter of the `ReturnTypeBuilderDelegate`? It's the same concept just more information is getting passed in.

Reply:
I merged those 2 methods. I also exposed a TypeSignatureSymbol, so it can be used in building signature. Currently, ReturnTypeBuilderDelegate with empty enumerable is called to get the type for signature purposes. Since my methods are more complicated than resourceGroup, managementGroup, etc. that are currently using this delegate, it seemed more reasonable to expose a dedicated property rather than rely on empty arguments list.